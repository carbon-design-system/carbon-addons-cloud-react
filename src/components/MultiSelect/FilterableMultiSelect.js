import cx from 'classnames';
import PropTypes from 'prop-types';
import React, { Fragment } from 'react';
import Downshift from 'downshift';
import isEqual from 'lodash.isequal';
import ListBox from '../ListBox';
import Checkbox from '../Checkbox';
import GroupLabel from '../GroupLabel';
import Selection from '../../internal/Selection';
import { sortingPropTypes } from './MultiSelectPropTypes';
import { defaultItemToString } from './tools/itemToString';
import { groupedByCategory } from './tools/groupedByCategory';
import { defaultSortItems, defaultCompareItems } from './tools/sorting';
import { defaultFilterItems } from '../ComboBox/tools/filter';

export default class FilterableMultiSelect extends React.Component {
  static propTypes = {
    ...sortingPropTypes,

    /**
     * Disable the control
     */
    disabled: PropTypes.bool,

    /**
     * We try to stay as generic as possible here to allow individuals to pass
     * in a collection of whatever kind of data structure they prefer
     */
    items: PropTypes.array.isRequired,

    /**
     * Allow users to pass in arbitrary items from their collection that are
     * pre-selected
     */
    initialSelectedItems: PropTypes.array,

    /**
     * Helper function passed to downshift that allows the library to render a
     * given item to a string label. By default, it extracts the `label` field
     * from a given item to serve as the item label in the list.
     */
    itemToString: PropTypes.func,

    /**
     * Specify the locale of the control. Used for the default `compareItems`
     * used for sorting the list of items in the control.
     */
    locale: PropTypes.string,

    /**
     * `onChange` is a utility for this controlled component to communicate to a
     * consuming component what kind of internal state changes are occuring.
     */
    onChange: PropTypes.func,

    /**
     * Generic `placeholder` that will be used as the textual representation of
     * what this field is for
     */
    placeholder: PropTypes.string.isRequired,

    /**
     * `true` to use the light version.
     */
    light: PropTypes.bool,
  };

  static defaultProps = {
    compareItems: defaultCompareItems,
    disabled: false,
    filterItems: defaultFilterItems,
    initialSelectedItems: [],
    itemToString: defaultItemToString,
    locale: 'en',
    sortItems: defaultSortItems,
    light: false,
  };

  constructor(props) {
    super(props);
    this.state = {
      highlightedIndex: null,
      isOpen: false,
      inputValue: '',
      openSections: [],
      checkedSuboptions: [],
      unCheckedSuboptions: [],
    };
  }

  handleOnChange = changes => {
    if (this.props.onChange) {
      this.props.onChange(changes);
    }
  };

  handleOnChangeSubOption = option => {
    if (!option.checked) {
      this.setState(prevState => ({
        checkedSuboptions: [...prevState.checkedSuboptions, option],
      }));
    } else {
      this.setState(prevState => ({
        checkedSuboptions: prevState.checkedSuboptions.filter(
          selectedOption => selectedOption !== option
        ),
      }));
    }
    option.checked = !option.checked;
  };

  onToggle = item => {
    !this.state.openSections.includes(item)
      ? this.setState({ openSections: [...this.state.openSections, item] })
      : this.setState(prevState => ({
          openSections: prevState.openSections.filter(
            itemOnState => itemOnState !== item
          ),
        }));
  };

  handleOnToggleMenu = () => {
    this.setState(state => ({
      isOpen: !state.isOpen,
    }));
  };

  handleOnOuterClick = () => {
    this.setState({
      isOpen: false,
    });
  };

  handleOnStateChange = changes => {
    const { type } = changes;
    switch (type) {
      case Downshift.stateChangeTypes.changeInput:
        this.setState({ inputValue: changes.inputValue });
        break;
      case Downshift.stateChangeTypes.keyDownArrowDown:
      case Downshift.stateChangeTypes.keyDownArrowUp:
      case Downshift.stateChangeTypes.itemMouseEnter:
        this.setState({ highlightedIndex: changes.highlightedIndex });
        break;
      case Downshift.stateChangeTypes.keyDownEscape:
      case Downshift.stateChangeTypes.mouseUp:
        this.setState({ isOpen: false });
        break;
      // Opt-in to some cases where we should be toggling the menu based on
      // a given key press or mouse handler
      // Reference: https://github.com/paypal/downshift/issues/206
      case Downshift.stateChangeTypes.clickButton:
      case Downshift.stateChangeTypes.keyDownSpaceButton:
        this.setState(() => {
          let nextIsOpen = changes.isOpen || false;
          if (changes.isOpen === false) {
            // If Downshift is trying to close the menu, but we know the input
            // is the active element in thedocument, then keep the menu open
            if (this.inputNode === document.activeElement) {
              nextIsOpen = true;
            }
          }
          return {
            isOpen: nextIsOpen,
          };
        });
        break;
    }
  };

  handleOnInputKeyDown = event => {
    event.stopPropagation();
  };

  handleOnInputValueChange = inputValue => {
    this.setState(() => {
      if (Array.isArray(inputValue)) {
        return {
          inputValue: '',
        };
      }
      return {
        inputValue: inputValue || '',
      };
    });
  };

  clearInputValue = event => {
    event.stopPropagation();
    this.setState({ inputValue: '' });
    this.inputNode && this.inputNode.focus && this.inputNode.focus();
  };

  handleSelectSubOptions = supOptions => {
    supOptions.map(option => {
      this.handleOnChangeSubOption(option);
    });
  };

  render() {
    const {
      highlightedIndex,
      isOpen,
      inputValue,
      checkedSuboptions,
      unCheckedSuboptions,
      openSections,
    } = this.state;
    const {
      className: containerClassName,
      disabled,
      filterItems,
      items,
      itemToString,
      initialSelectedItems,
      id,
      locale,
      placeholder,
      sortItems,
      compareItems,
      light,
    } = this.props;
    const className = cx(
      'bx--multi-select',
      'bx--combo-box',
      containerClassName,
      {
        'bx--list-box--light': light,
      }
    );
    return (
      <Selection
        onChange={this.handleOnChange}
        initialSelectedItems={initialSelectedItems}
        render={({ selectedItems, onItemChange, clearSelection }) => (
          <Downshift
            highlightedIndex={highlightedIndex}
            isOpen={isOpen}
            inputValue={inputValue}
            onInputValueChange={this.handleOnInputValueChange}
            itemToString={itemToString}
            onStateChange={this.handleOnStateChange}
            onOuterClick={this.handleOnOuterClick}
            selectedItem={selectedItems}
            render={({
              getButtonProps,
              getInputProps,
              getItemProps,
              getRootProps,
              isOpen,
              inputValue,
              selectedItem,
            }) => (
              <ListBox
                style={{ outline: 'none' }}
                className={className}
                disabled={disabled}
                {...getRootProps({ refKey: 'innerRef' })}>
                <ListBox.Field {...getButtonProps({ disabled })}>
                  {selectedItem.length > 0 && (
                    <ListBox.Selection
                      clearSelection={e => {
                        {
                          clearSelection(e);
                        }
                      }}
                      selectionCount={selectedItem.length}
                    />
                  )}
                  <input
                    className="bx--text-input"
                    ref={el => (this.inputNode = el)}
                    {...getInputProps({
                      disabled,
                      id,
                      placeholder,
                      onKeyDown: this.handleOnInputKeyDown,
                    })}
                  />
                  {inputValue && isOpen && (
                    <ListBox.Selection clearSelection={this.clearInputValue} />
                  )}
                  <ListBox.MenuIcon isOpen={isOpen} />
                </ListBox.Field>
                {isOpen && (
                  <ListBox.Menu style={{ maxHeight: '424px' }}>
                    {groupedByCategory(items).map(group => {
                      const hasGroups = group[0] !== 'undefined' ? true : false;
                      let categoryName = '';
                      hasGroups
                        ? (categoryName = group[0].toUpperCase())
                        : null;
                      return (
                        <Fragment>
                          {hasGroups && (
                            <div>
                              <GroupLabel>{categoryName}</GroupLabel>
                            </div>
                          )}
                          {sortItems(
                            filterItems(group[1], { itemToString, inputValue }),
                            {
                              selectedItems,
                              itemToString,
                              compareItems,
                              locale,
                            }
                          ).map(item => {
                            const itemProps = getItemProps({ item });
                            const itemText = itemToString(item);
                            const isChecked =
                              selectedItem.filter(selected =>
                                isEqual(selected, item)
                              ).length > 0;
                            const subOptions = item.options;
                            const groupIsOpen =
                              openSections.filter(groupOpen =>
                                isEqual(groupOpen, item)
                              ).length > 0;
                            const myCheckedOptions = item.options.filter(
                              subOption => subOption.checked == true
                            );
                            const myUncheckedOptions = item.options.filter(
                              subOption => subOption.checked != true
                            );
                            const currentParent = item;
                            return (
                              <Fragment>
                                <ListBox.MenuItem
                                  key={itemProps.id}
                                  isActive={isChecked}
                                  onClick={e => {
                                    {
                                      const clickOutOfCheckBox =
                                        e.target.localName != 'label';
                                      if (clickOutOfCheckBox) {
                                        this.onToggle(item);
                                      } else {
                                        onItemChange(item);
                                        debugger;
                                        if (
                                          myCheckedOptions.length == 0 &&
                                          !selectedItems.includes(item)
                                        ) {
                                          !groupIsOpen
                                            ? this.onToggle(item)
                                            : null;
                                          this.handleSelectSubOptions(
                                            myUncheckedOptions
                                          );
                                        } else {
                                          this.handleSelectSubOptions(
                                            myCheckedOptions
                                          );
                                        }
                                      }
                                    }
                                  }}>
                                  <Checkbox
                                    id={itemProps.id}
                                    name={itemText}
                                    checked={isChecked}
                                    readOnly={true}
                                    tabIndex="-1"
                                    labelText={itemText}
                                    hasGroups={hasGroups}
                                    isExpanded={groupIsOpen}
                                  />
                                </ListBox.MenuItem>

                                {groupIsOpen &&
                                  subOptions != undefined &&
                                  subOptions.map((item, index) => {
                                    const optionsProps = getItemProps({ item });
                                    const isCheckedSub =
                                      this.state.checkedSuboptions.filter(
                                        selected => isEqual(selected, item)
                                      ).length > 0;
                                    const subOpText = itemToString(item);
                                    const checkBoxIndex = index.toString();
                                    return (
                                      <ListBox.MenuItem
                                        key={optionsProps.id}
                                        style={{ paddingLeft: '35px' }}
                                        isActive={isCheckedSub}
                                        onClick={e => {
                                          {
                                            const onlySupOpChecked =
                                              myCheckedOptions.length == 1 &&
                                              myCheckedOptions.includes(item);
                                            onlySupOpChecked ||
                                            myCheckedOptions.length == 0
                                              ? onItemChange(currentParent)
                                              : null;
                                            this.handleOnChangeSubOption(item);
                                          }
                                        }}>
                                        <Checkbox
                                          id={checkBoxIndex}
                                          name={subOpText}
                                          checked={isCheckedSub}
                                          labelText={subOpText}
                                          readOnly={true}
                                          tabIndex="-1"
                                        />
                                      </ListBox.MenuItem>
                                    );
                                  })}
                              </Fragment>
                            );
                          })}
                        </Fragment>
                      );
                    })}
                  </ListBox.Menu>
                )}
              </ListBox>
            )}
          />
        )}
      />
    );
  }
}
