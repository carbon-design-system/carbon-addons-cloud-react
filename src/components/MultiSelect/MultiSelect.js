import cx from 'classnames';
import React from 'react';
import PropTypes from 'prop-types';
import Downshift from 'downshift';
import isEqual from 'lodash.isequal';
import ListBox from '../ListBox';
import Checkbox from '../Checkbox';
import Selection from '../../internal/Selection';
import { sortingPropTypes } from './MultiSelectPropTypes';
import { defaultItemToString } from './tools/itemToString';
import { defaultSortItems, defaultCompareItems } from './tools/sorting';

export default class MultiSelect extends React.Component {
  static propTypes = {
    ...sortingPropTypes,

    /**
     * Disable the control
     */
    disabled: PropTypes.bool,

    /**
     * We try to stay as generic as possible here to allow individuals to pass
     * in a collection of whatever kind of data structure they prefer
     */
    items: PropTypes.array.isRequired,

    /**
     * Allow users to pass in arbitrary items from their collection that are
     * pre-selected
     */
    initialSelectedItems: PropTypes.array,

    /**
     * Helper function passed to downshift that allows the library to render a
     * given item to a string label. By default, it extracts the `label` field
     * from a given item to serve as the item label in the list.
     */
    itemToString: PropTypes.func,

    /**
     * Generic `label` that will be used as the textual representation of what
     * this field is for
     */
    label: PropTypes.node.isRequired,

    /**
     * Specify the locale of the control. Used for the default `compareItems`
     * used for sorting the list of items in the control.
     */
    locale: PropTypes.string,

    /**
     * `onChange` is a utility for this controlled component to communicate to a
     * consuming component what kind of internal state changes are occuring.
     */
    onChange: PropTypes.func,
    /**
     * Specify 'inline' to create an inline multi-select.
     */
    type: PropTypes.oneOf(['default', 'inline']),

    /**
     *  Specify title to show title on hover
     */
    useTitleInItem: PropTypes.bool,

    /**
     * `true` to use the light version.
     */
    light: PropTypes.bool,
    /**
     * Is the current selection invalid?
     */
    invalid: PropTypes.bool,
    /**
     * If invalid, what is the error?
     */
    invalidText: PropTypes.string,
  };

  static defaultProps = {
    compareItems: defaultCompareItems,
    disabled: false,
    locale: 'en',
    itemToString: defaultItemToString,
    initialSelectedItems: [],
    sortItems: defaultSortItems,
    type: 'default',
    light: false,
    title: false,
  };

  constructor(props) {
    super(props);
    this.state = {
      highlightedIndex: null,
      isOpen: false,
    };
  }

  handleOnChange = changes => {
    if (this.props.onChange) {
      this.props.onChange(changes);
    }
  };

  handleOnOuterClick = () => {
    this.setState({
      isOpen: false,
    });
  };

  handleOnStateChange = changes => {
    const { type } = changes;
    switch (type) {
      case Downshift.stateChangeTypes.keyDownArrowDown:
      case Downshift.stateChangeTypes.keyDownArrowUp:
      case Downshift.stateChangeTypes.itemMouseEnter:
        this.setState({ highlightedIndex: changes.highlightedIndex });
        break;
      case Downshift.stateChangeTypes.keyDownEscape:
      case Downshift.stateChangeTypes.mouseUp:
        this.setState({ isOpen: false });
        break;
      // Opt-in to some cases where we should be toggling the menu based on
      // a given key press or mouse handler
      // Reference: https://github.com/paypal/downshift/issues/206
      case Downshift.stateChangeTypes.clickButton:
      case Downshift.stateChangeTypes.keyDownSpaceButton:
        this.setState(() => {
          let nextIsOpen = changes.isOpen || false;
          if (changes.isOpen === false) {
            // If Downshift is trying to close the menu, but we know the input
            // is the active element in the document, then keep the menu open
            if (this.inputNode === document.activeElement) {
              nextIsOpen = true;
            }
          }
          return {
            isOpen: nextIsOpen,
          };
        });
        break;
    }
  };

  render() {
    const { highlightedIndex, isOpen } = this.state;
    const {
      className: containerClassName,
      items,
      itemToString,
      label,
      type,
      disabled,
      initialSelectedItems,
      sortItems,
      compareItems,
      light,
      invalid,
      invalidText,
      useTitleInItem,
    } = this.props;
    const className = cx('bx--multi-select', containerClassName, {
      'bx--list-box--light': light,
    });
    return (
      <Selection
        onChange={this.handleOnChange}
        initialSelectedItems={initialSelectedItems}
        render={({ selectedItems, onItemChange, clearSelection }) => (
          <Downshift
            highlightedIndex={highlightedIndex}
            isOpen={isOpen}
            itemToString={itemToString}
            onChange={onItemChange}
            onStateChange={this.handleOnStateChange}
            onOuterClick={this.handleOnOuterClick}
            selectedItem={selectedItems}
            render={({
              getRootProps,
              selectedItem,
              isOpen,
              itemToString,
              highlightedIndex,
              getItemProps,
              getButtonProps,
            }) => (
              <ListBox
                type={type}
                className={className}
                disabled={disabled}
                invalid={invalid}
                invalidText={invalidText}
                {...getRootProps({ refKey: 'innerRef' })}>
                <ListBox.Field {...getButtonProps({ disabled })}>
                  {selectedItem.length > 0 && (
                    <ListBox.Selection
                      clearSelection={clearSelection}
                      selectionCount={selectedItem.length}
                    />
                  )}
                  <span className="bx--list-box__label">{label}</span>
                  <ListBox.MenuIcon isOpen={isOpen} />
                </ListBox.Field>
                {isOpen && (
                  <ListBox.Menu>
                    {sortItems(items, {
                      selectedItems,
                      itemToString,
                      compareItems,
                      locale: 'en',
                    }).map((item, index) => {
                      const itemProps = getItemProps({ item });
                      const itemText = itemToString(item);
                      const isChecked =
                        selectedItem.filter(selected => isEqual(selected, item))
                          .length > 0;
                      return (
                        <ListBox.MenuItem
                          key={itemProps.id}
                          isActive={isChecked}
                          isHighlighted={highlightedIndex === index}
                          {...itemProps}>
                          <Checkbox
                            id={itemProps.id}
                            title={useTitleInItem ? itemText : null}
                            name={itemText}
                            checked={isChecked}
                            readOnly={true}
                            tabIndex="-1"
                            labelText={itemText}
                          />
                        </ListBox.MenuItem>
                      );
                    })}
                  </ListBox.Menu>
                )}
              </ListBox>
            )}
          />
        )}
      />
    );
  }
}
