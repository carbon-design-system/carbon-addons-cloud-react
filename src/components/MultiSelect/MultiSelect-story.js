import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';

import { withKnobs, boolean, select, text } from '@storybook/addon-knobs';
import MultiSelect from '../MultiSelect';

const items = [
  {
    id: 'item-1',
    text: 'Item 1',
  },
  {
    id: 'item-2',
    text: 'Item 2',
  },
];

const defaultLabel = 'MultiSelect Label';
const defaultPlaceholder = 'Filter';

const types = {
  'Default (default)': 'default',
  'Inline (inline)': 'inline',
};

const props = () => ({
  filterable: boolean(
    'Filterable (`<MultiSelect.Filterable>` instead of `<MultiSelect>`)',
    false
  ),
  disabled: boolean('Disabled (disabled)', false),
  light: boolean('Light variant (light)', false),
  useTitleInItem: boolean('Show tooltip on hover', false),
  type: select('UI type (Only for `<MultiSelect>`) (type)', types, 'default'),
  label: text('Label (label)', defaultLabel),
  invalid: boolean('Show form validation UI (invalid)', false),
  invalidText: text(
    'Form validation UI content (invalidText)',
    'Invalid Selection'
  ),
  onChange: action('onChange'),
});

storiesOf('MultiSelect', module)
  .addDecorator(withKnobs)
  .add(
    'default',
    () => {
      const { filterable, ...multiSelectProps } = props();
      const ComponentToUse = !filterable ? MultiSelect : MultiSelect.Filterable;
      const placeholder = !filterable ? undefined : defaultPlaceholder;
      return (
        <div style={{ width: 300 }}>
          <ComponentToUse
            {...multiSelectProps}
            items={items}
            itemToString={item => (item ? item.text : '')}
            placeholder={placeholder}
          />
        </div>
      );
    },
    {
      info: {
        text: `
            MultiSelect
          `,
      },
    }
  )
  .add(
    'with initial selected items',
    () => {
      const { filterable, ...multiSelectProps } = props();
      const ComponentToUse = !filterable ? MultiSelect : MultiSelect.Filterable;
      const placeholder = !filterable ? undefined : defaultPlaceholder;
      return (
        <div style={{ width: 300 }}>
          <ComponentToUse
            {...multiSelectProps}
            items={items}
            itemToString={item => (item ? item.text : '')}
            initialSelectedItems={[items[0], items[1]]}
            placeholder={placeholder}
          />
        </div>
      );
    },
    {
      info: {
        text: `
            Provide a set of items to initially select in the control
          `,
      },
    }
  );
