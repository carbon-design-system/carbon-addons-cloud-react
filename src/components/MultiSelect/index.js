import MultiSelect from './MultiSelect';
import FilterableMultiSelect from './FilterableMultiSelect';

MultiSelect.Filterable = FilterableMultiSelect;

export default MultiSelect;
