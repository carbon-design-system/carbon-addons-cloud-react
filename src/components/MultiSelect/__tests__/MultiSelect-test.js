import React from 'react';
import { mount } from 'enzyme';
import MultiSelect from '../../MultiSelect';
import {
  openMenu,
  generateItems,
  generateGenericItem,
} from '../../ListBox/test-helpers';

const mouseDownAndUp = node => {
  node.dispatchEvent(new window.MouseEvent('mousedown', { bubbles: true }));
  node.dispatchEvent(new window.MouseEvent('mouseup', { bubbles: true }));
};

describe('MultiSelect', () => {
  it('should render', () => {
    const wrapper = mount(
      <MultiSelect
        label="Field"
        items={generateItems(5, generateGenericItem)}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should initialize with no selected items if no `initialSelectedItems` are given', () => {
    const items = generateItems(5, generateGenericItem);
    const wrapper = mount(<MultiSelect label="Field" items={items} />);
    expect(wrapper.find('Selection').instance().state.selectedItems).toEqual(
      []
    );
  });
  it('should initialize with the menu not open', () => {
    const items = generateItems(5, generateGenericItem);
    const wrapper = mount(<MultiSelect label="Field" items={items} />);
    expect(wrapper.state('isOpen')).toEqual(false);
  });

  describe('#handleOnToggleMenu', () => {
    it('should toggle the boolean `isOpen` field', () => {
      const items = generateItems(5, generateGenericItem);
      const wrapper = mount(<MultiSelect label="Field" items={items} />);
      expect(wrapper.state('isOpen')).toBe(false);
      wrapper.find('.bx--list-box__field').simulate('click');
      expect(wrapper.state('isOpen')).toBe(true);
      wrapper.find('.bx--list-box__field').simulate('click');
      expect(wrapper.state('isOpen')).toBe(false);
    });
  });

  describe('when `initialSelectedItems` is given', () => {
    it('should initialize `selectedItems` with the given initial selected items', () => {
      const items = generateItems(5, generateGenericItem);
      const wrapper = mount(
        <MultiSelect
          label="Field"
          items={items}
          initialSelectedItems={[items[0], items[1]]}
        />
      );
      expect(wrapper.find('Selection').instance().state.selectedItems).toEqual([
        items[0],
        items[1],
      ]);
    });
  });

  describe('MultiSelect with InitialSelectedItems', () => {
    let mockProps;
    const items = generateItems(5, generateGenericItem);

    beforeEach(() => {
      mockProps = {
        items: items,
        initialSelectedItems: [items[0], items[1], items[2]],
        itemToString: ({ label }) => label,
        onChange: jest.fn(),
        label: 'Label',
      };
    });

    it('should allow a user to de-select an item by clicking on initial selected items', () => {
      const wrapper = mount(<MultiSelect {...mockProps} />);
      expect(
        wrapper.find('Selection').instance().state.selectedItems.length
      ).toBe(3);

      wrapper.find('.bx--list-box__field').simulate('click');
      wrapper
        .find('.bx--list-box__menu-item')
        .at(0)
        .simulate('click');

      expect(
        wrapper.find('Selection').instance().state.selectedItems.length
      ).toBe(2);
    });

    it('should allow a user to de-select an initial selected item by hitting enter on initial selected item', () => {
      const wrapper = mount(<MultiSelect {...mockProps} />);
      const simulateArrowDown = wrapper =>
        wrapper.find('.bx--list-box__field').simulate('keydown', {
          key: 'ArrowDown',
        });

      expect(
        wrapper.find('Selection').instance().state.selectedItems.length
      ).toBe(3);
      openMenu(wrapper);
      simulateArrowDown(wrapper);
      wrapper
        .find('.bx--list-box__field')
        .at(0)
        .simulate('keydown', {
          key: 'Enter',
        });
      expect(
        wrapper.find('Selection').instance().state.selectedItems.length
      ).toBe(2);
    });

    it('should select an item when a user clicks on an item', () => {
      const wrapper = mount(<MultiSelect {...mockProps} />);
      expect(
        wrapper.find('Selection').instance().state.selectedItems.length
      ).toBe(3);

      wrapper.find('.bx--list-box__field').simulate('click');
      wrapper
        .find('.bx--list-box__menu-item')
        .at(4)
        .simulate('click');

      expect(
        wrapper.find('Selection').instance().state.selectedItems.length
      ).toBe(4);
    });
  });

  describe('e2e', () => {
    let mockProps;

    beforeEach(() => {
      mockProps = {
        items: generateItems(5, generateGenericItem),
        initialSelectedItems: [],
        itemToString: ({ label }) => label,
        onChange: jest.fn(),
        label: 'Label',
      };
    });

    it('should open the menu when a user clicks on the ListBox field', () => {
      const wrapper = mount(<MultiSelect {...mockProps} />);
      wrapper.find('.bx--list-box__field').simulate('click');
      expect(wrapper.find('.bx--list-box__menu').length).toBe(1);
      expect(wrapper.find('.bx--list-box__menu-item').length).toBe(
        mockProps.items.length
      );
    });

    it('should open the menu when a user focuses and hits space on the ListBox field', () => {
      const wrapper = mount(<MultiSelect {...mockProps} />);
      wrapper.find('.bx--list-box__field').simulate('keydown', {
        key: ' ',
      });
      expect(wrapper.find('.bx--list-box__menu').length).toBe(1);
      expect(wrapper.find('.bx--list-box__menu-item').length).toBe(
        mockProps.items.length
      );
    });

    it('should select an item when a user clicks on an item', () => {
      const wrapper = mount(<MultiSelect {...mockProps} />);
      openMenu(wrapper);
      wrapper
        .find('.bx--list-box__menu-item')
        .first()
        .simulate('click');
      expect(wrapper.find('Selection').instance().state.selectedItems).toEqual([
        mockProps.items[0],
      ]);
    });

    it('should allow a user to highlight items with the up and down arrow keys', () => {
      const wrapper = mount(<MultiSelect {...mockProps} />);
      wrapper.find('.bx--list-box__field').simulate('click');
      const simulateArrowDown = () =>
        wrapper.find('.bx--list-box__field').simulate('keydown', {
          key: 'ArrowDown',
        });
      const simulateArrowUp = () =>
        wrapper.find('.bx--list-box__field').simulate('keydown', {
          key: 'ArrowUp',
        });
      const getHighlightedId = () =>
        wrapper.find('.bx--list-box__menu-item--highlighted').prop('id');
      simulateArrowDown();
      expect(getHighlightedId()).toBe('downshift-12-item-0');
      simulateArrowDown();
      expect(getHighlightedId()).toBe('downshift-12-item-1');
      // Simulate "wrap" behavior
      simulateArrowDown();
      simulateArrowDown();
      simulateArrowDown();
      simulateArrowDown();
      expect(getHighlightedId()).toBe('downshift-12-item-0');
      simulateArrowUp();
      expect(getHighlightedId()).toBe('downshift-12-item-4');
    });

    it('should close the menu when a user clicks outside of the control', () => {
      const wrapper = mount(<MultiSelect {...mockProps} />);
      wrapper.find('.bx--list-box__field').simulate('click');
      mouseDownAndUp(document.body);
      expect(wrapper.state('isOpen')).toBe(false);
    });

    it('should show a badge that mirrors the number of selected items', () => {
      const wrapper = mount(<MultiSelect {...mockProps} />);
      openMenu(wrapper);
      for (let i = 0; i < mockProps.items.length; i++) {
        wrapper
          .find('.bx--list-box__menu-item')
          .at(i)
          .simulate('click');
        expect(wrapper.find('.bx--list-box__selection').text()).toEqual(
          expect.stringContaining(`${i + 1}`)
        );
      }
      expect(wrapper.find('Selection').instance().state.selectedItems).toEqual(
        mockProps.items
      );
    });

    it('should allow a user to de-select an item by clicking on a selected item', () => {
      const wrapper = mount(<MultiSelect {...mockProps} />);
      wrapper.find('.bx--list-box__field').simulate('click');
      wrapper
        .find('.bx--list-box__menu-item')
        .at(0)
        .simulate('click');
      expect(wrapper.find('.bx--list-box__menu-item--active').length).toBe(1);
      wrapper
        .find('.bx--list-box__menu-item')
        .at(0)
        .simulate('click');
      expect(wrapper.find('.bx--list-box__menu-item--active').length).toBe(0);
    });

    it('should allow a user to de-select an item by hitting enter on a selected item', () => {
      const wrapper = mount(<MultiSelect {...mockProps} />);
      const simulateArrowDown = wrapper =>
        wrapper.find('.bx--list-box__field').simulate('keydown', {
          key: 'ArrowDown',
        });
      openMenu(wrapper);
      wrapper
        .find('.bx--list-box__menu-item')
        .at(0)
        .simulate('click');
      expect(wrapper.find('.bx--list-box__menu-item--active').length).toBe(1);
      simulateArrowDown(wrapper);
      wrapper.find('.bx--list-box__field').simulate('keydown', {
        key: 'Enter',
      });
      expect(wrapper.find('.bx--list-box__menu-item--active').length).toBe(0);
    });

    it('should allow a user to click on the clear icon to clear all selected items', () => {
      const wrapper = mount(<MultiSelect {...mockProps} />);
      openMenu(wrapper);
      wrapper
        .find('.bx--list-box__menu-item')
        .at(0)
        .simulate('click');
      wrapper.find('.bx--list-box__selection').simulate('click');
      expect(wrapper.find('Selection').instance().state.selectedItems).toEqual(
        []
      );
    });

    it('should allow a user to focus the clear icon and hit enter to clear all selected items', () => {
      const wrapper = mount(<MultiSelect {...mockProps} />);
      openMenu(wrapper);
      wrapper
        .find('.bx--list-box__menu-item')
        .at(0)
        .simulate('click');
      wrapper.find('.bx--list-box__selection').simulate('keydown', {
        keyCode: 13,
      });
      expect(wrapper.find('Selection').instance().state.selectedItems).toEqual(
        []
      );
    });
  });
});
