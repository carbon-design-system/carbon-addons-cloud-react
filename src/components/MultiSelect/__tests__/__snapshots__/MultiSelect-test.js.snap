// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiSelect should render 1`] = `
<MultiSelect
  compareItems={[Function]}
  disabled={false}
  initialSelectedItems={Array []}
  itemToString={[Function]}
  items={
    Array [
      Object {
        "id": "id-0",
        "label": "Item 0",
        "value": 0,
      },
      Object {
        "id": "id-1",
        "label": "Item 1",
        "value": 1,
      },
      Object {
        "id": "id-2",
        "label": "Item 2",
        "value": 2,
      },
      Object {
        "id": "id-3",
        "label": "Item 3",
        "value": 3,
      },
      Object {
        "id": "id-4",
        "label": "Item 4",
        "value": 4,
      },
    ]
  }
  label="Field"
  light={false}
  locale="en"
  sortItems={[Function]}
  title={false}
  type="default"
>
  <Selection
    initialSelectedItems={Array []}
    onChange={[Function]}
    render={[Function]}
  >
    <Downshift
      breakingChanges={Object {}}
      defaultHighlightedIndex={null}
      defaultInputValue=""
      defaultIsOpen={false}
      defaultSelectedItem={null}
      environment={[Window]}
      getA11yStatusMessage={[Function]}
      highlightedIndex={null}
      isOpen={false}
      itemToString={[Function]}
      onChange={[Function]}
      onInputValueChange={[Function]}
      onOuterClick={[Function]}
      onSelect={[Function]}
      onStateChange={[Function]}
      onUserAction={[Function]}
      render={[Function]}
      selectedItem={Array []}
      selectedItemChanged={[Function]}
      stateReducer={[Function]}
    >
      <ListBox
        ariaLabel="Choose an item"
        className="bx--multi-select"
        disabled={false}
        innerRef={[Function]}
        type="default"
      >
        <div
          aria-label="Choose an item"
          className="bx--multi-select bx--list-box"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="listbox"
          tabIndex="0"
        >
          <ListBoxField
            aria-expanded={false}
            aria-haspopup={true}
            aria-label="open menu"
            data-toggle={true}
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            type="button"
          >
            <div
              aria-expanded={false}
              aria-haspopup={true}
              aria-label="open menu"
              className="bx--list-box__field"
              data-toggle={true}
              disabled={false}
              onBlur={[Function]}
              onClick={[Function]}
              onKeyDown={[Function]}
              role="button"
              tabIndex="0"
              type="button"
            >
              <span
                className="bx--list-box__label"
              >
                Field
              </span>
              <ListBoxMenuIcon
                isOpen={false}
                translateWithId={[Function]}
              >
                <div
                  className="bx--list-box__menu-icon"
                >
                  <Icon
                    alt="Open menu"
                    description="Open menu"
                    fillRule="evenodd"
                    icon={
                      Object {
                        "height": "5",
                        "id": "icon--caret--down",
                        "name": "icon--caret--down",
                        "styles": "",
                        "svgData": Object {
                          "circles": "",
                          "ellipses": "",
                          "paths": Array [
                            Object {
                              "d": "M0 0l5 4.998L10 0z",
                            },
                          ],
                          "polygons": "",
                          "polylines": "",
                          "rects": "",
                        },
                        "tags": "icon--caret--down",
                        "viewBox": "0 0 10 5",
                        "width": "10",
                      }
                    }
                    role="img"
                  >
                    <svg
                      alt="Open menu"
                      aria-label="Open menu"
                      fillRule="evenodd"
                      height="5"
                      role="img"
                      viewBox="0 0 10 5"
                      width="10"
                    >
                      <title>
                        Open menu
                      </title>
                      <path
                        d="M0 0l5 4.998L10 0z"
                        key="key0"
                      />
                    </svg>
                  </Icon>
                </div>
              </ListBoxMenuIcon>
            </div>
          </ListBoxField>
        </div>
      </ListBox>
    </Downshift>
  </Selection>
</MultiSelect>
`;
